CoreGraphFramework/test/unit/src/au/gov/asd/tac/constellation/graph/attribute/io/AbstractCachedStringIOProviderNGTest.java

```diff
@@ -0,0 +1,159 @@
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyInt;
+import static org.mockito.ArgumentMatchers.anyObject;
+import org.mockito.Mockito;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.when;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+/**
+ *
+ * @author serpens24
+ */
+public class AbstractCachedStringIOProviderNGTest {
+        
+    // Create object under test
+    StringIOProvider instance = new StringIOProvider();
+
+    // Define mocks
+    GraphReadMethods mockGraphReadMethods;
+    GraphWriteMethods mockGraphWriteMethods;
+    JsonNode mockJsonNode;
+    JsonGenerator mockJsonGenerator;
+    ImmutableObjectCache mockCache;
+        
+    // Test variables
+    final int attributeId = 23;
+    final int elementId = 41;
+    final String attribValue = "TestAttrib";
+    final String decoratorStr = "TestDecorator";
+    final GraphAttribute attr = new GraphAttribute(attributeId, GraphElementType.GRAPH, "attrType", "attrName", "attrDesc", null, null);
+ 
+    
+    public AbstractCachedStringIOProviderNGTest() {
+    }
+
+    @BeforeClass
+    public static void setUpClass() throws Exception {
+    }
+
+    @AfterClass
+    public static void tearDownClass() throws Exception {
+    }
+
+    @BeforeMethod
+    public void setUpMethod() throws Exception {
+        instance = new StringIOProvider();
+    }
+
+    @AfterMethod
+    public void tearDownMethod() throws Exception {
+    }
+    
+    /**
+     * Perform reset of all mocks and argument captors to ensure clean test steps.
+     */
+    public void resetMocking() {
+        mockGraphReadMethods = mock(GraphReadMethods.class);
+        mockGraphWriteMethods = mock(GraphWriteMethods.class);
+        mockJsonNode = mock(JsonNode.class);
+        mockJsonGenerator = mock(JsonGenerator.class);
+        mockCache = mock(ImmutableObjectCache.class);
+    }
+
+    /**
+     * Test of readObject method, of class AbstractCachedStringIOProvider.
+     */
+    @Test
+    public void testReadObject() throws Exception {
+        System.out.println("AbstractCachedStringIOProviderNGTest.testReadObject");
+
+        // Call method under test with JsonNode set to returen isNull = true
+        resetMocking();
+        when(mockJsonNode.isNull()).thenReturn(true);
+        when(mockCache.deduplicate(anyObject())).thenReturn(attribValue);
+        instance.readObject(attributeId, elementId, mockJsonNode, mockGraphWriteMethods, null, null, null, mockCache);
+        Mockito.verify(mockJsonNode, times(0)).textValue();
+        Mockito.verify(mockCache, times(1)).deduplicate(null);
+
+        // Call method under test with JsonNode set to returen isNull = false
+        resetMocking();
```

Comment:
could be worth separating use cases into their own tests if you're needing to reset each time (resetting is then done in `setUpMethod`). Otherwise it's worth noting that `thenReturn` can be specified to return different values for each call. e.g. `when(myMock.myFunction()).thenReturn(true, false)` will return `true` the first time the function is called on the mock and `false` the 2nd time

Reply:
Thanks Antares, I spoke to @arcturus2 today about splitting the tests out or not and we came ot the conclusion he was happy leaving as is for now .... BUT the thenReturn usage was something I really need on a test I'm currently working on and had 'just' finished finding out yu could do it on google... this will make my life a LOT easier